src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -380,5 +405,26 @@ private static string GetDisplayGroup(FieldInfo field)
 
             return string.Empty;
         }
+
+        // Internal for testing
+        internal static bool IsNonNullable(IEnumerable<object> attributes)
```

Comment:
Note the almost identical logic about to go into EF Core: https://github.com/aspnet/EntityFrameworkCore/pull/15499/files#diff-0ebd43bbbecf10e14f32e65b224af763R28

I actually cache some type information about NullableAttribute for better perf, while supporting multiple NullableAttribute (since it may be synthesized several times into different participating assemblies...).

Reply:
@roji does you code need to be thread-safe? That appears to not be...

Reply:
Nope, the "convention" you are seeing is a composable piece of code that runs once at startup, during EF Core's model building, and instances aren't supposed to be thread-safe. However, we've seen cases of EF Core models which were so big that the model building process was quite slow, which is why I'm doing this caching.